feat(core): rewrite input and output docs (#21269)
nrwl/nx: docs/shared/recipes/running-tasks/customizing-inputs.md

---

> title: Fine-tuning Caching with Inputs

Reviewer: This document isn't a recipe any more - it was border line before and has clearly crossed the line.  It's kind of hybrid between a reference doc and a concept page, but we should move it into the concept section if we're not going to re-write it completely as separate docs.

description: 'Learn how to optimize when Nx restores from the cache by customizing what should be taken into consideration when calculating computation hashes for tasks'
---

# Fine-tuning Caching with Inputs

When Nx [computes the hash for a given operation](/concepts/how-caching-works), it takes into account the `inputs` of the target. The `inputs` are a list of **file sets**, **runtime** inputs and **environment variables** that affect the output of the target. If any of the `inputs` change, the cache is invalidated and the target is re-run.

First, let's take a look at the [different types of inputs](#types-of-inputs) that Nx uses to compute the computation hash of a task and how to [define common sets of inputs](#named-inputs) used by many tasks.
Then, the guide will explore how to [view the inputs of a task](#viewing-the-inputs-of-a-task) in your workspace and how to [configure inputs](#configuring-inputs) for a task.
At the end, there will be some examples of common inputs configurations which many workspaces may use.

> Throughout this recipe, the following project structure of a simple workspace will be used as an example to help understand inputs better.

Author: The projects are referenced now

{% /graph %}

## Types of Inputs

Nx can consider the following types of Inputs when computing the hash for a given operation.

### Project Configuration

The configuration of a project includes the definition of the task itself. Changing this will often result in a change of behavior of running different tasks.

Nx always considers the configuration of the project of a task and its dependencies when calculating the hash of a task.

### Command Arguments

> Passing different arguments to a Nx command will often result in a change of behavior of running tasks. Nx will filter out arguments which are for Nx itself and do not have any effect on running of tasks such as `--parallel` or `--projects`. Nx will always take into consideration command arguments when calculating the hash for a task.

Reviewer: ```suggestion
Passing different arguments to a Nx command will often result in a task changing behavior. Because of this, Nx will take into consideration command arguments when calculating the hash for a task. Nx will filter out arguments which are for Nx itself and do not have any effect on the actual tasks such as `--parallel` or `--projects`.
```

> For example, running `nx build myreactapp --prod` will not reuse the same results of running `nx build myreactapp` earlier.

Reviewer: ```suggestion
For example, running `nx build myreactapp --prod` will not reuse the cached output of `nx build myreactapp`.
```

### Source Files

> Changing source code will often result in a change of behavior of running different tasks. Nx can consider the contents of files matching a pattern when calculating the computation hash.

Reviewer: ```suggestion
Changing source code will often result in a task changing behavior. Nx can consider the contents of files matching a pattern when calculating the computation hash.
```

Source File inputs are defined as such:

```jsonc
"inputs": {
  "{projectRoot}/**/*", // All files in a project
  "{workspaceRoot}/.gitignore", // A specific file in the workspace
  "{projectRoot}/**/*.ts", // A glob pattern for files
  "!{projectRoot}/**/*.spec.ts", // Excluding files matching a glob pattern
}
```

Source File inputs must be prefixed with either `{projectRoot}` or `{workspaceRoot}` to distinguish where the paths should be resolved from.

Prefixing a Source File input with `!` will exclude the files matching the pattern from the set of files used to calculate the hash.

By default, Nx will use all files in a project as well as all files in the project's dependencies when computing a hash for tasks belonging to the project.
This may cause Nx to rerun some tasks even when files irrelevant to the task have changed but it ensures that by default, Nx always re-runs the task when it should.

See the [Common Input Sets section](#common-inputs) to see some common examples of source file inputs

### Environment Variables

> Tools and scripts will often use some Environment Variables to change their behavior. Nx can hash the value of environment variables when it runs in the cache hash key calculation in order to invalidate the cache if the environment variable value changes. Environment Variable inputs are defined as such:

Reviewer: ```suggestion
Tools and scripts will often use some Environment Variables to change their behavior. Nx can hash the value of environment variables when it runs the cache hash key calculation. Environment Variable inputs are defined as such:
```

```jsonc
"inputs": [
  { "env": "API_KEY" } // this will include the value of $API_KEY in the cache hash
]
```

### Runtime Inputs

> You can use a Runtime input to provide a script which will output the information you want to include in the computation hash. Runtime inputs are defined as such:

Reviewer: ```suggestion
You can use a Runtime input to provide a script which will output the information you want to include in the computation hash. Runtime inputs are defined like this:
```

```jsonc
"inputs": [
  { "runtime": "node --version" }
]
```

> This kind of input is often used to include versions of tools used to run the task. An advanced use case for this type of input is to provide a custom script which will have its output considered in the hash.

Reviewer: ```suggestion
This kind of input is often used to include versions of tools used to run the task.
```

### External Dependencies

> Source code often imports from external dependencies from installed through package managers. For example, a React application will likely import `react`. Nx will always consider External Dependencies depended upon by any source code within the project.

Reviewer: ```suggestion
Source code often imports from external dependencies installed through package managers. For example, a React application will likely import `react`. Nx will always consider External Dependencies depended upon by any source code within the project.
```

> External Dependencies are also often tools used to run tasks. Updating the versions of those tools will change the behavior of those tasks. External Dependencies inputs are defined as such:

Reviewer 2: We should make it clearer that the external dependencies are only effective if you're running a task without an executor, i.e., just the command option or with a plugin that implicitly creates targets. It does not affect the dependencies of the actual project

> External Dependencies are also often tools used to run tasks. Updating the versions of those tools will change the behavior of those tasks. External Dependencies inputs are defined as such:

Reviewer: ```suggestion
External Dependencies are also often tools used to run tasks. Updating the versions of those tools will change the behavior of those tasks. External Dependencies inputs are defined like this:
```

```jsonc
"inputs": [
  { "externalDependencies": ["jest"] }
]
```

By default, if no external dependencies inputs are specified, Nx will include the hash of all external dependencies in the computation hash.
For many targets, Nx does not know which External Dependencies are used and which are not.
By considering all External Dependencies, Nx will always re-run tasks when necessary even though in some cases, it might have been able to restore from cache.

For example, when defining a target to run `eslint .`, Nx does not know which ESLint Plugins are used by the eslint config.
To be safe, it will consider all external dependencies, including all ESLint plugins, to ensure that the `lint` task is re-run when any of the plugins have been updated.
The drawback of this assumption is that Nx will re-run the `lint` task even when external dependencies not used by ESLint are updated.

```jsonc
"targets": {
  "lint": {
    "command": "eslint ."
  }
}
```

> This default behavior can be overridden by adding any external dependency inputs and will enable Nx to use its cache to be used more.

Reviewer: ```suggestion
This default behavior can be overridden by adding any external dependency inputs which enables the Nx cache to be used more.
```

```jsonc {% highlightLines=[6] %}
"targets": {
  "lint": {
    "command": "eslint .",
    "inputs": [
      "default",
      { "externalDependencies": ["eslint", "eslint-config-airbnb"] }
    ]
  }
}
```

> Targets which use an Executor of Nx Plugins maintained by the Nx Team will have the correct set of external dependencies considered when calculating the hash.

Reviewer: ```suggestion
Targets which use an Executor from an Nx plugin maintained by the Nx Team will have the correct set of external dependencies considered when calculating the hash.
```

]
```

The pattern will be matched with outputs of tasks which this task depends on. Setting `transitive` to `true` will also include outputs of all task dependencies of this task in the [task pipeline](/concepts/task-pipeline-configuration).

### A subset of the root `tsconfig.json` or `tsconfig.base.json`

When a root `tsconfig.json` or `tsconfig.base.json` is present, Nx will always consider parts of the file which apply to the project of a task being run.
This includes the full `compilerOptions` section and particular path mappings in the `compilerOptions.paths` property.
This allows Nx to to not invalidate every single task when a path mapping is added or removed from the root `tsconfig.json` file

## Named Inputs

Many tasks will utilize the same sets of inputs with minor differences. Nx allows you to define these sets in the `namedInputs` section of the `nx.json` file.

These sets of inputs can then be referenced and reused in the `inputs` array of targets. You can think of these as variables which can be used in the `inputs` array.

Named Inputs can be defined for the entire workspace in `nx.json` or for a specific project in `project.json`.

> Once defined for the workspace, any tasks for all projects in the workspace can use the named inputs in the `inputs` array.

Reviewer 2: We should mention that Nx will always include a `"default": "{projectRoot}/**/*"` named input, even if its not specified. But it will be overridden if it is specified.

> {% tabs %}

Reviewer: The code in these tabs doesn't look right.  There may have been some copy-paste problems.

Reviewer: I think these tabs should be sections with headers, not tabs.  People aren't choosing which path they want and ignoring the rest, they need to learn about all of them.  Tabs are for scenarios when the audience only cares about one option of many.

{% tab label="Default Named Inputs" %}

By default, Nx Workspaces are generated with the following named inputs:

```jsonc {% fileName="nx.json" highlightLines=["2-6"] %}
{
  "namedInputs": {
    "default": ["{projectRoot}/**/*", "sharedGlobals"], // Default Inputs

>     "production": ["default", "!{projectRoot}/some-dev-only-files"], // Production Inputs

Reviewer: This doesn't look like defaults.

    "sharedGlobals": [] // Shared Global Inputs
  },
  "targetDefaults": {
    "inputs": ["default", "^default"]
  }
}
```

{% /tab %}
{% tab label="Workspace (nx.json)" %}

The following example shows how to define named inputs for the entire workspace in `nx.json`.

```jsonc {% fileName="nx.json" highlightLines=["3-6"] %}
{
  "namedInputs": {
    "default": ["{projectRoot}/**/*"] // Default Inputs
  }
}
```

{% /tab %}
{% tab label="Project Level (project.json)" %}

```jsonc {% fileName="nx.json" highlightLines=["2-6"] %}
{
  "namedInputs": {
    "default": ["{projectRoot}/**/*", "sharedGlobals"], // Default Inputs
    "production": ["default", "!{projectRoot}/some-dev-only-files"], // Production Inputs
    "sharedGlobals": [] // Shared Global Inputs
  },
  "targetDefaults": {
    "inputs": ["default", "^default"]
  }
}
```

{% /tab %}
{% tab %}
Naming a set of inputs with the same name as a set of inputs for the workspace will override the meaning of that set for the project.
{% /tab %}
{% /tabs %}

> ### Default Inputs

Reviewer: Default, Shared Global and Production inputs should be under the "Default Named inputs section" so it is clear they are only default names and can be overriden

The `default` inputs include all files in a project as well as any shared global inputs which should be considered for all tasks.

### Shared Global Inputs

The `sharedGlobal` inputs include things that Nx should always look at when determining computation hashes. For instance, this could be the OS where the command is being run or the version of Node.

### Production Inputs

The `production` inputs only include the set of files in a project which will affect behavior of the project in production.
For instance, the `main.ts` of an application will be compiled as the main application logic and will affect the end user experience while files such as `.eslintrc.json` only affect the tools used by the developers and have no direct impact on the end user experience.

> In general, it is best to define the `production` inputs as the `default` inputs (everything) excluding the specific files which are known not to affect end user experience. This makes it so that by default, all files are considered to affect end user behavior unless excluded from the `production` fileset.

Reviewer 2: I think this needs to be clear that its just explaining the default generated named inputs. I know we mention it in the first tab here:
<img width="932" alt="image" src="https://github.com/nrwl/nx/assets/4332460/370677fe-75b0-40cc-ac2d-45f08fc35e60">
but as soon as you switch tabs, you lose that context. I got confused because I saw this:
<img width="969" alt="image" src="https://github.com/nrwl/nx/assets/4332460/bbede1b7-e20c-4207-a357-7b75b259ccc1">

The headings of each of the "namedInput" names make it seem more important that it is.

```shell
nx build myreactapp --graph
```

This will show the task graph executed by Nx when running the command.

Clicking the task will open a tooltip which lists out all of the inputs of the task. A button within the tooltip will also reveal more details about the configuration for the project which the task belongs to.

Use this tool to help understand what inputs are being used by Nx to determine when the cache can be used.

{% callout title="Note" type="info" %}
If no `inputs` are specified at all, Nx will default to looking at all files of a project and its dependencies. This is a rather cautious approach. This might cause Nx to re-run a task in some cases where the cache could have been used instead but it will always give you correct output.
{% /callout %}

## Configuring Inputs

Inputs of a task are configured in the `inputs` array on the target. This can be done in several places:

- Nx Plugins often infer inputs for tasks which run other tools.
- The `inputs` array can be configured in the `targetDefaults` for a set of targets.
- The `inputs` array can be configured for a specific target in the project configuration file.

> {% tabs %}

Reviewer: These tabs should be headings

{% tab label="Nx Plugins" %}
Nx plugins specified in the `plugins` array in `nx.json` will define inputs for the tasks which it infers to be cacheable.

For example, the `@nx/webpack/plugin` plugin, will add the following inputs for tasks which use `webpack` to compile:

```jsonc
  "inputs": [
    "production", // Production source siles of the project

>     "^production", // Production source files of dependencies

Reviewer 2: This might be relevant outside this PR, but if plugins automatically infer using the `production` namedInput, and it doesnt exist in `nx.json` - we will throw an error. 

https://github.com/nrwl/nx/blob/3333ef49684c2f1540ba63408bcc942c99426043/packages/nx/src/native/tasks/inputs.rs#L201-L210

    { "externalDependencies": ["webpack"] } // The version of `webpack` used.
  ]

> ```

Reviewer 2: This will be confusing for newcomers because they wouldn't know where this is. (This is automatically created and can only be viewed when doing `nx show project <projectName> --web`)

Author: I have now added the project details view so they can see it instead of my having the snippet of config readers won't see.
